torch/utils/_python_dispatch.py

```diff
@@ -151,7 +151,99 @@ def transform_subclass(t, callback):
+               # - TensorImpl.set_sizes_strides() therefore yells if you try to directly change sizes.
+               torch.ops.aten.set_.source_Storage_storage_offset(ret, arg.untyped_storage(), 0, ret.shape)
+           else:
+               ret.set_(arg.untyped_storage())
+
+    def is_match(arg, ret):
+        arg_aliases = set() if not arg.alias_info else arg.alias_info.before_set
+        out_aliases = set() if not ret.alias_info else ret.alias_info.before_set
+        return len(arg_aliases & out_aliases) > 0
+
+    num_args = len(func._schema.arguments)
+    num_returns = len(func._schema.returns)
+    for arg_idx in range(num_args):
+        for return_idx in range(num_returns):
+            if is_match(func._schema.arguments[arg_idx], func._schema.returns[return_idx]):
+                alias_storage(args[arg_idx], outs[return_idx])
+
+    # Sigh... the torchscript parser has a bug where alias annotations for Tensor[](a) don't show up properly
+    # See https://github.com/pytorch/pytorch/issues/106173
+    if func.overloadpacket in [
+        torch.ops.aten.chunk,
+        torch.ops.aten.tensor_split,
+        torch.ops.aten.split,
+        torch.ops.aten.split_with_sizes,
+        torch.ops.aten.hsplit,
+        torch.ops.aten.vsplit,
+        torch.ops.aten.dsplit,
+        torch.ops.aten.unbind,
+    ]:
+        assert isinstance(outs, list) and all(isinstance(x, torch.Tensor) for x in outs)
+        for o in outs:
+            # For lists of outputs, need to alias every individual tensor to the input
+            alias_storage(args[0], o)
+
+def return_and_correct_aliasing(func, args, out):
+    def get_write_alias(x):
+        if not x.alias_info or not x.alias_info.before_set:
+            return None
+        before_set = list(x.alias_info.before_set)
+        # torchscript allows for complicated alias sets, but our dispatcher ops only really involve simple aliasing
+        assert len(before_set) == 1
+        if '!' in list(before_set)[0]:
+            return before_set[0]
+        return None
+
+    # Fix up the storages of any outs so that they point to the same storage as the input,
+    # if func is a view op.
+    _correct_storage_aliasing(func, args, [out] if not isinstance(out, (list, tuple)) else out)
+
+    # Next: we need to make sure to return inputs directly, if the output is a mutable alias (e.g. add_()).
+
+    # simple case: none of our outputs have mutable aliases, so we can return the output as-is
+    if not any(get_write_alias(r) is not None for r in func._schema.returns):
+        return out
+
+    # simplifying assumption: we don't have **any** ops with return types like "-> (Tensor(a!), Tensor)"
+    if not all(get_write_alias(r) is not None for r in func._schema.returns):
+        raise RuntimeError("Unsupported schema: " + str(func._schema))
+
+    if len(func._schema.returns) == 1:
+        arg_idx = get_arg_idx_from_alias(get_write_alias(0))
+        return args[arg_idx]
```

Comment:
Let's say we have `sin_(Tensor(a!) x) -> Tensor(a!)`. Is the contract that the output must be the exact same as the input, or is it fine if the output is a view of the input (but we have written to both?)

I'm trying to figure out if we return the Tensor as-is is because we understand `sin_` to be an in-place operation, or if this information is encoded in the schema, or both.